Allow collection of EnsoGL stats when Monitor panel is not visible

akavel opened the pull request:
### Pull Request Description

<!--
- Please describe the nature of your PR here, as well as the motivation for it.
- If it fixes an open issue, please mention that issue number here.
-->

This change makes EnsoGL runtime stats be always collected, even when EnsoGL `Monitor` panel is not visible. Those stats are intended to be used in the future by a profiling framework.

**Performance impact:** Continuous collection of stats introduces an overhead of two Web Performance API `now()` calls in each frame of the main rendering loop, plus a small number of simple arithmetic calculations. This is assumed to be a negligible and acceptable overhead.

#### Visuals



A screenshot of the Monitor panel in full `ide` after applying the PR, taken in IDE built with `./run dist`:


<img width="991" alt="Screenshot 2022-02-14 at 16 11 42" src="https://user-images.githubusercontent.com/273837/153891378-8a2fb333-34ce-46ce-99df-7d796817310c.png">

A recording, also in IDE built with `./run dist`; note that FPS is impacted by the act of recording itself:





https://user-images.githubusercontent.com/273837/154104016-49a12e23-1210-4477-9743-ec1611e5b4ed.mov




https://www.pivotaltracker.com/story/show/181093601

### Important Notes

<!--
- Mention important elements of the design.
- Mention any notable changes to APIs.
-->

 - Responsibility for controlling how `Stats` gathering and calculation is performed at various points in the main rendering loop was removed from `Monitor` - the `Monitor`'s purpose is only to display existing data, it should not influence how the data is collected.
 - Two previously existing distinct `Monitor` structs were merged into one, to avoid confusion; after previous refactorings, the remaining `stats::Monitor` did not have much useful code anyway.
 - In `stats` package, refactoring was done, to make `StatsData` a "dumb", data-only type, and to move the logic related to stats collection and frame tracking to other helper types.


[ci no changelog needed]

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guides.
- [x] All documentation and configuration conforms to the [markdown](https://github.com/enso-org/enso/blob/develop/docs/style-guide/markdown.md) and [YAML](https://github.com/enso-org/enso/blob/develop/docs/style-guide/yaml.md) style guides.
- [x] All code has been tested where possible.

akavel committed:
add FrameTime test

akavel committed:
add Fps test

akavel committed:
add Fps test showing ValueCheck::Error

akavel committed:
test FrameTime warning & error threshold

akavel committed:
move calculations from Samplers to Stats

akavel committed:
adjust tests to new code

akavel committed:
bring back FrameTime & WasmMemory structs

akavel committed:
add some section comments

akavel committed:
use cast::AsPrimitive instead of disabling lint warning

akavel committed:
Merge remote-tracking branch 'origin/develop' into wip/akavel/ensogl-stats-vs-sampler-181093700

akavel committed:
dummy commit to force refresh no-changelog CI check

akavel committed:
Merge branch 'develop' into wip/akavel/ensogl-stats-vs-sampler-181093700

akavel committed:
initial raw port of relevant parts of old PR

akavel committed:
review: rename PanelData::end()

akavel committed:
make FrameTime and WasmMemory use stats_sampler! macro

akavel committed:
review: make tests less repetitive

akavel committed:
review: reduce repetition in tests with macro

akavel committed:
review: improve comment in test, describing influence of delay on FPS

akavel committed:
review: add comment in begin_frame

akavel committed:
Merge branch 'develop' into wip/akavel/ensogl-stats-vs-sampler-181093700

akavel committed:
simplify the werid name TestSamplerByFrame

akavel committed:
begin_frame -> StatsData

akavel committed:
WIP sample previous frame stats

akavel committed:
refactor

akavel committed:
various FIXMEs

akavel committed:
add max_value to stats_sampler!

akavel committed:
WIP removing new()

akavel committed:
review: add a TODO about `cfg!(wasm32)`

akavel committed:
delete check_by_threshold() helper

akavel committed:
rename macro args and inline struct def

akavel committed:
StatsData doc for struct, disable docs for fields

akavel committed:
allow missing docs only on pub fields

akavel committed:
delete seemingly unnecessary frame_counter()

akavel committed:
WIP updating tests

akavel committed:
fix tests & frame 0 checks

akavel committed:
cargo fmt

akavel committed:
update tests comments

akavel committed:
improve some comments

akavel committed:
Merge remote-tracking branch 'origin/develop' into wip/akavel/fix-fps-sampler-181140499

akavel committed:
fix clippy

akavel committed:
bring back frame_counter(), cannot remove it yet

akavel committed:
tweak comments and names

akavel committed:
reduce usage of allow(missing_docs)

akavel committed:
add overlooked Copy traits

akavel committed:
first sketch of a test; adding pub for fields

akavel committed:
more test

akavel committed:
starting to use macro in tests to reduce repetitiveness

akavel committed:
convert further tests to macro

akavel committed:
ValueAccumulator fields seem to not need to be pub

akavel committed:
drop serde for now

akavel committed:
improve readability in test

akavel committed:
small tweak to comments

akavel committed:
Merge branch 'develop' into wip/akavel/stats-summaries-181093920

akavel committed:
add and call on_stats_available registry

akavel committed:
add callback registration function

akavel committed:
add missing newline

akavel committed:
review: WIP start adding new demo scene "stats"

akavel committed:
WIP show simple label

akavel committed:
WIP modify label text in loop

akavel committed:
WIP show FPS

akavel committed:
WIP show fps summary in demo scene

akavel committed:
WIP update FPS only once per 60 frames to speed things up

akavel committed:
tweak TODOs in demo scene

akavel committed:
rename things in demo scene

akavel committed:
review: make macro pattern multiline

akavel committed:
improve var names and delete comments in macro

akavel committed:
Merge remote-tracking branch 'origin/develop' into wip/akavel/stats-summaries-181093920

akavel committed:
calculate always, but print only every 60 frames

akavel committed:
add missing "the"

akavel committed:
Merge remote-tracking branch 'origin/develop' into wip/akavel/on-stats-available-181093832

akavel committed:
Merge branch 'wip/akavel/stats-summaries-181093920' into wip/akavel/on-stats-available-181093832

akavel committed:
first draft of on_stats_available usage

akavel committed:
skip 0th frame

akavel committed:
bring back FPS display

akavel committed:
move vars

akavel committed:
delete unused begin_value field in PanelData

akavel committed:
adjust comment on Monitor::begin()

akavel committed:
don't return StatsData if prev frame was incomplete

akavel committed:
move vars to local scope

akavel committed:
Revert "move vars to local scope"

This reverts commit a9887f0be72553d313d53660375a3e9274f7fb69.

akavel committed:
reverse inner condition

akavel committed:
reverse outer condition

akavel committed:
WIP in demo scene, always collect stats

akavel committed:
Merge remote-tracking branch 'origin/develop' into wip/akavel/on-stats-available-181093832

akavel committed:
cleanup a None

akavel committed:
are_stats_valid + comment

akavel committed:
s/collection/calculations

akavel committed:
Merge branch 'develop' into wip/akavel/on-stats-available-181093832

akavel committed:
review: WIP frame_measurement status tracking

akavel committed:
improve comment on previous_frame_stats

akavel committed:
move enum further down

akavel committed:
fix some English style in comment

akavel committed:
simplify comment for previous_frame_stats

akavel committed:
Merge remote-tracking branch 'origin/develop' into wip/akavel/on-stats-available-181093832

akavel committed:
small improvement

akavel committed:
s/FrameMeasurement/...State/

akavel committed:
extra comment

akavel committed:
s/.../frame_measurement_state

akavel committed:
s/prev/previous

akavel committed:
further renames

akavel committed:
use stats_enabled()

akavel committed:
rename stats_collection_enabled()

akavel committed:
add missing comment

akavel committed:
extra comment

wdanilo reviewed (changes requested):
I don't understand this design and thus I don't like it - why this is implemented as a "switch" in the World struct? Who needs it? What is the desired usage scenario? Why the World struct is responsible for remembering it? Are we sure that this is the desired design? I believe these questions are **impossibe** to be answered without seeing a code that requires this functionality – maybe the responsibilities should belong to another entity?

This is another example of PR which implements dead code (similar to the previous one). I am not sure if this is the right design. I would like to see the usage of this functionality in the profiling framework first in order to understand if this should be implemented this way or other way around. I'm not sure whether this should be implemented as a "switch" in the World struct, or maybe profiling framework should manage it somehow else. And it's impossible to decide without seeing the code in context of its usage.

I would like all of our PRs to never be implemented as a dead code (like this PR is), always with a clear usage scenario. The design of the code should be dictated by how the thing is intended to be used. Right now, we have several super small changes with new structures and APIs not used by anyone and IMO there is a high probability that after the final PR will be created with the profiling framework, I will request changes to the API implemented in these PRs, because the API should be judged by the usage scenario. So I'd love ask to close this PR and implement PRs with code changes that are required by real required usage, not artificial debug scenes.

MichaelMauderer commented on lib/rust/ensogl/core/src/display/world/stats.rs:
```diff
@@ -53,22 +55,36 @@ impl {
+    /// Makes the [`Monitor`] run [`Stats`] measurements even when the Monitor panel is not
+    /// [`visible()`]. (By default, measurements are only done when [`visible()`], as an
+    /// optimization.)
+    pub fn always_collect_stats(&mut self) {
```
Should there not also be a switch to disable this?

akavel committed:
just always collect stats

akavel committed:
move stats calculations outside of Monitor

akavel committed:
typo

akavel committed:
purge no longer used accessor

akavel committed:
another typo

akavel committed:
WIP merging two Monitors

akavel committed:
fix to compile

akavel committed:
delete empty module

akavel committed:
move stats_monitor.draw() to on_stats_available

akavel committed:
small cleanup

akavel committed:
cargo fmt

akavel committed:
add comments

akavel committed:
small tweaks

akavel committed:
s/plots/graphs/ for uniformness

akavel committed:
revert order in main_loop capture

akavel commented:
@wdanilo I refactored the code as inspired by our discussions on chat (the video & screenshot are also updated); PTAL if it makes more sense now. Thanks for all the valuable lessons!

akavel committed:
Merge branch 'develop' into wip/akavel/on-stats-available-181093832

akavel committed:
fix clippy

akavel committed:
delete stats::{Accumulator,Summary} and "stats" demo scene

MichaelMauderer commented on lib/rust/ensogl/core/src/debug/monitor.rs:
```diff
@@ -226,35 +278,25 @@ impl Default for Monitor {
 
-    /// Check whether the mointor is visible.
-    pub fn visible(&self) -> bool {
+    fn panels(&self) -> &Vec<Panel> {
```
Could this have the return type `&[Panel]`?

akavel committed:
Revert "delete stats::{Accumulator,Summary} and "stats" demo scene"

This reverts commit 3161901ee6a86986f983fef9fab5d6f9071cf47f.

akavel committed:
review: change panels() return value

akavel committed:
Merge branch 'develop' into wip/akavel/on-stats-available-181093832

wdanilo commented on lib/rust/ensogl/core/src/display/world.rs:
```diff
@@ -76,33 +76,35 @@ impl World {
                 // Note [Main Loop Performance]
 
-                stats_monitor.begin();
+                let previous_frame_stats = stats.begin_frame(performance.now());
```
Is this world's responsibility to measure the time and provide the current time value to stats collector? Why world is handling that, but for example resetting the stats per frame is handled by stats collector?

kazcw replied:
I think this inconsistency is because `begin_frame` is misplaced. It is an operation for building up a sequence of `StatsData` values; it seems odd for sequence logic to be in a method on a `StatsData` object.

wdanilo commented on lib/rust/ensogl/core/src/debug/monitor.rs:
```diff
@@ -198,9 +198,64 @@ impl Drop for DomData {
+
+    /// Draw the monitor and update its graphs based on the provided stats data.
+    pub fn draw(&self, stats: &StatsData) {
+        let mut renderer = self.renderer.borrow_mut();
```
This is a bad pattern to borrow_mut for the whole scope. Someone will refactor this code in the future, put her borrows in the lines below and it will blow up. Better way of doing it is the way that function `toggle` is implemented - borrow once and redirect the call to the borrowed structure.

wdanilo commented on lib/rust/ensogl/core/src/debug/stats.rs:
```diff
@@ -210,6 +205,7 @@ impl StatsData {
         };
         self.frame_counter += 1;
```
I was writing about it in another PR review – I do not understand why this field is here. Also, I think its name is very confusing. This is frame counter of only these frames where the stats were gathered, this is not "all frames counter", as the name may suggest.

Moreover, it seems that the only place this variable is being used is when it is compared to 0 to check if this is the first sample, so why we are collecting information of rendered frames if no one uses this information?

wdanilo replied:
Also, no docs describe that stats collector assumes that it is running in every frame – that was explained on Discord, but this information is missing from the code.

akavel replied:
**Re: `frame_counter`:** I changed the field to a boolean flag, thanks! Other than that, I'm not fully sure if what you were commenting on is just that it was a number, or that it exists at all? If the latter, then I don't currently see good alternatives:

 - I could add an `is_default()` method to `StatsData`, checking that every field equals zero, but that would be much more complex and also less effective;
 - I could add an `is_valid()` method checking e.g. just some fields (`self.fps != 0.0 || self.frame_counter != 0.0`), but I'm not convinced that would be more clear; an `initialized` field (or whatever name we'll agree upon) sounds most direct as to intention;
 - I can still name the field differently than `initialized`; however, e.g. `is_first_sample` doesn't sound appropriate to me, as this is notably not the *first* frame/sample that is "invalid", but the "0-th" - i.e., "the predecessor to the 1st frame" - and naming it `is_frame_zero` also sounds confusing to me, esp. given that programmers at various times count from 1 or from 0, so it's still not clear if 0 is "first" or not... other than that, I can't think of any other good names for it - I initially thought of `valid`, but I find it arguable which one is more clear.

kazcw replied:
What about `Option<StatsData>`?

akavel replied:
~~@kazcw Initially (some time ago) I suggested to return `Option<StatsData>` from `begin_frame()`, but Wojciech then suggested to rather return `default()` than `None` in case of the "uninitialized" frame. Would your suggestion be to track this as `Option<StatsData>` inside, but then return `default()` from `begin_frame()` instead of `None`? Or to return `None` from `begin_frame()` in that case?~~

I modified the code to return `Option<StatsData>` from `begin_frame()`, as I believe this is indeed a better design.

akavel committed:
review: avoid long-scoped borrow_mut()

akavel committed:
review: add a note in docs about missed frames

akavel committed:
review: WIP simplifying frame_counter

akavel committed:
Merge remote-tracking branch 'origin/develop' into wip/akavel/on-stats-available-181093832

akavel committed:
Merge branch 'wip/akavel/on-stats-available-181093832' into wip/akavel/allow-continuous-stats-181093601

akavel committed:
s/valid_frame/valid

akavel committed:
move valid=true assignment

akavel committed:
s/valid/initialized

akavel committed:
WIP TimeProvider

akavel committed:
WIP FrameStats

akavel committed:
WIP fix Stats to compile

akavel committed:
WIP make stuff compile

akavel committed:
fix first warnings + cargo fmt

akavel committed:
tweak comments & visibility

akavel committed:
subsection formatting

akavel committed:
comment tweak

akavel committed:
tweak comments, names, visibility

akavel committed:
CHORE rebuild updated package-lock

akavel committed:
drop T:Clone requirement

akavel committed:
MockTimeProvider

akavel committed:
fix tests

akavel committed:
cargo fmt

akavel committed:
delete obsolete field

akavel commented:
@wdanilo @kazcw PTAL; inspired by a chat with Kaz on Friday, I modified the code to what I believe is a better design.

akavel committed:
Merge remote-tracking branch 'origin/develop' into wip/akavel/on-stats-available-181093832

wdanilo commented on lib/rust/ensogl/core/src/debug/monitor.rs:
```diff
@@ -786,33 +817,82 @@ stats_sampler!(
+                let mismatch_msg = iformat!("expected: " $expected_check;? ", got: " tested_check;?);
+                assert!(matches!(tested_check, $expected_check), "{}", mismatch_msg);
+            } else {
+                assert!(false, "Got None from begin_frame()");
```
Errors should be valid English sentences. Please apply to other errors too.

wdanilo commented on lib/rust/ensogl/core/src/debug/monitor.rs:
```diff
@@ -823,10 +903,8 @@ mod tests {
-        // will always be 0.0 and "Correct" at this point (which is a special case - depending on
-        // tested stat, for later frames 0.0 could result in a threshold warning/error).
-        test_and_advance_frame!(test, 0.0, ValueCheck::Correct; next: 10.0, 6.0);
+        // Note: there is no frame before "Frame 1", so the previous frame will not be tested.
```
What do you mean by "there is no frame before "Frame 1""? Before the first rendering frame the program can evaluate expressions, so there is time before frame 1.

MichaelMauderer replied:
Needs reworking of the comment and/or additional check for `None` value.

wdanilo commented on lib/rust/ensogl/core/src/debug/stats.rs:
```diff
@@ -15,48 +15,137 @@
 
 
 
+// ====================
+// === TimeProvider ===
+// ====================
+
+/// An entity that can retrieve current time.
+pub trait TimeProvider {
+    /// Returns current time, measured in milliseconds.
+    fn now(&self) -> f64;
+}
+
+impl TimeProvider for Performance {
+    fn now(&self) -> f64 {
+        self.now()
+    }
+}
```
This should be refactored to other abstractions over web-functions (I dont remember the module where we keep them).

wdanilo commented on lib/rust/ensogl/core/src/debug/stats.rs:
```diff
@@ -15,48 +15,137 @@
+    }
+
+    fn begin_frame(&mut self) -> Option<StatsData> {
+        let time = self.time_provider.now();
+        let new_frame = StatsData {
+            draw_call_count: 0,
+            shader_compile_count: 0,
+            data_upload_count: 0,
+            data_upload_size: 0,
+            ..self.stats_data
+        };
+        let mut previous_frame = mem::replace(&mut self.stats_data, new_frame);
+        match self.frame_begin_time.replace(time) {
+            Some(previous_frame_begin_time) => {
+                let end_time = time;
+                previous_frame.fps = 1000.0 / (end_time - previous_frame_begin_time);
+                Some(previous_frame)
+            }
+            None => None,
+        }
+    }
```
this function is huge. The previous design with `reset_per_frame_statistics` was nicer because it made such functions easier to read.

wdanilo commented on lib/rust/ensogl/core/src/debug/stats.rs:
```diff
@@ -15,48 +15,137 @@
+
+    fn begin_frame(&mut self) -> Option<StatsData> {
+        let time = self.time_provider.now();
+        let new_frame = StatsData {
+            draw_call_count: 0,
+            shader_compile_count: 0,
+            data_upload_count: 0,
+            data_upload_size: 0,
+            ..self.stats_data
+        };
```
the var naming is strange here. This var does not contain a new frame, it's misleading.

wdanilo commented on lib/rust/ensogl/core/src/debug/stats.rs:
```diff
@@ -15,48 +15,137 @@
+            ..self.stats_data
+        };
+        let mut previous_frame = mem::replace(&mut self.stats_data, new_frame);
+        match self.frame_begin_time.replace(time) {
```
this can be implemented as a mapping for Option, which we always prefer instead of writing `None => None`.

wdanilo commented on lib/rust/ensogl/core/src/debug/stats.rs:
```diff
@@ -15,48 +15,137 @@
+// ======================
+
+#[derive(Debug)]
+struct StatsCollector<T: TimeProvider> {
+    time_provider:    T,
+    stats_data:       StatsData,
+    frame_begin_time: Option<f64>,
+}
```
I'm not sure whether this struct is collecting stats. It has a reference to StatsData (which contains stats) and then it cleans the stats on every frame, and adds info about memory usage and frame time after the frame. I would not name it a "stats collector" in such a case.

MichaelMauderer replied:
Alternatives `FrameStatsData`.

wdanilo commented on lib/rust/ensogl/core/src/debug/stats.rs:
```diff
@@ -15,48 +15,137 @@
+    ///
+    /// Note: on first ever frame, there was no "previous frame", so no data will be returned.
+    ///
+    /// Note: the code works under an assumption that [`begin_frame()`] and [`end_frame()`] are
+    /// called properly on every frame (behavior in case of missed frames or missed calls is not
+    /// specified).
```
The wording "are called properly" is not explained. What does "properly" mean?

wdanilo commented on lib/rust/ensogl/core/src/debug/stats.rs:
```diff
@@ -15,48 +15,137 @@
+/// Tracks stats related to the current rendering frame, providing methods for modifying and
+/// retrieving their values.
+#[derive(Debug, CloneRef)]
+pub struct StatsOverTime<T: TimeProvider> {
+    rc: Rc<RefCell<StatsCollector<T>>>,
```
I don't understand why a "stats collector" which is put in Rc<RefCell<...>> is named `StatsOverTime`. So a shared stats collector are stats over time?

MichaelMauderer replied:
Alternatives `SharedStatsHandler`. (Or `Shared`...whatever the inner struct is).

wdanilo commented on lib/rust/ensogl/core/src/debug/stats.rs:
```diff
@@ -15,48 +15,137 @@
+
+// === StatsOverTime ===
+
+/// Tracks stats related to the current rendering frame, providing methods for modifying and
+/// retrieving their values.
```
What does mean "stats related to the current frame"?

wdanilo commented on lib/rust/ensogl/core/src/debug/monitor.rs:
```diff
@@ -786,33 +817,82 @@ stats_sampler!(
+    // === MockTimeProvider ===
+
+    #[derive(Default, Clone)]
+    struct MockTimeProvider {
+        t: Rc<RefCell<f64>>,
+    }
+
+    impl TimeProvider for MockTimeProvider {
+        fn now(&self) -> f64 {
+            *self.t.borrow()
+        }
+    }
+
+    impl AddAssign<f64> for MockTimeProvider {
+        fn add_assign(&mut self, dt: f64) {
+            *self.t.borrow_mut() += dt;
+        }
+    }
```
This is a nice design!

wdanilo commented on lib/rust/ensogl/core/src/debug/monitor.rs:
```diff
@@ -786,33 +817,82 @@ stats_sampler!(
+            if let Some(prev_frame_stats) = prev_frame_stats {
+                let tested_value = $test.sampler.value(&prev_frame_stats);
+                let tested_check = $test.sampler.check(&prev_frame_stats);
+                assert_approx_eq!(tested_value, $expected_value, 0.001);
```
please extract all constants, like this `0.001` to well named variables explaining what they are.

wdanilo commented on lib/rust/ensogl/core/src/debug/stats.rs:
```diff
@@ -15,48 +15,137 @@
+/// Tracks stats related to the current rendering frame, providing methods for modifying and
+/// retrieving their values.
+#[derive(Debug, CloneRef)]
+pub struct StatsOverTime<T: TimeProvider> {
```
Please do not use constraints in structs definition. I was explaining it multiple times in different PRs. Such constraints leak everywhere. For example, you implement later `Clone` for this structure. Is it really necessary that `T:TimeProvider` when cloning the structure?

There are big topics about it on Haskell mailing lists and it is considered a big anti-pattern in Haskell (which is supposed to be banned). I believe that we should ban it too.

akavel committed:
Merge remote-tracking branch 'origin/develop' into wip/akavel/on-stats-available-181093832

akavel committed:
Merge branch 'wip/akavel/on-stats-available-181093832' into wip/akavel/allow-continuous-stats-181093601

akavel committed:
review: move TimeProvider to enso_web crate

akavel committed:
review: drop constraints from struct definitions

akavel committed:
review: bring back reset_per_frame_stats

akavel committed:
fix import in test

akavel committed:
review: use const in test

akavel committed:
review: make error messages be full sentences

akavel committed:
review: rename types as discussed in video call

akavel committed:
some comments

akavel committed:
review: clarify test scenario contents

akavel committed:
review: adjust comment in test

akavel committed:
review: adjust comment for begin_frame()

akavel committed:
review: edit comments for StatsWithTimeProvider

akavel committed:
Merge remote-tracking branch 'origin/develop' into wip/akavel/allow-continuous-stats-181093601

akavel committed:
delete constraint in struct definition

akavel committed:
improve language in error msg

akavel committed:
improve language further

akavel committed:
further improve language in test comments

akavel committed:
further improve language in doc comments

akavel committed:
further improve language in doc comments

akavel committed:
fix var name

akavel committed:
comment tweak

akavel committed:
Merge remote-tracking branch 'origin/develop' into wip/akavel/allow-continuous-stats-181093601

akavel committed:
Merge branch 'develop' into wip/akavel/allow-continuous-stats-181093601

akavel committed:
Merge branch 'develop' into wip/akavel/allow-continuous-stats-181093601

akavel commented:
(*Process note:* the code-review is now accepted, therefore the PR is waiting for Review by @MichaelMauderer on Task https://www.pivotaltracker.com/story/show/181093601 before merge.)

mergify[bot] committed:
Merge branch 'develop' into wip/akavel/allow-continuous-stats-181093601

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/enso-org/enso?pullRequest=3260) <br/>All committers have signed the CLA.

mergify[bot] committed:
Merge branch 'develop' into wip/akavel/allow-continuous-stats-181093601

mergify[bot] committed:
Merge branch 'develop' into wip/akavel/allow-continuous-stats-181093601
